src/EditorFeatures/Core/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,211 @@
+        private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+        private readonly ICodeDefinitionWindowService _codeDefinitionWindowService;
+
+        [ImportingConstructor]
+        public DefinitionContextConnectionListener(
+            IMetadataAsSourceFileService metadataAsSourceFileService,
+            ICodeDefinitionWindowService codeDefinitionWindowService)
+        {
+            _metadataAsSourceFileService = metadataAsSourceFileService;
+            _codeDefinitionWindowService = codeDefinitionWindowService;
+        }
+
+        void IWpfTextViewConnectionListener.SubjectBuffersConnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (!_views.ContainsKey(textView) && !textView.Roles.Contains(PredefinedTextViewRoles.CodeDefinitionView))
+            {
+                _views.Add(textView, new CancellationTokenSource());
+                textView.Caret.PositionChanged += OnTextViewCaretPositionChanged;
+                var fireAndForget = UpdateDefinitionContext(textView, textView.Caret.Position);
+            }
+        }
+
+        void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (reason == ConnectionReason.TextViewLifetime ||
+                !textView.BufferGraph.GetTextBuffers(b => b.ContentType.IsOfType(ContentTypeNames.RoslynContentType)).Any())
+            {
+                CancellationTokenSource cancellationTokenSource;
+                if (_views.TryGetValue(textView, out cancellationTokenSource))
+                {
+                    cancellationTokenSource.Cancel();
+                    _views.Remove(textView);
+                    textView.Caret.PositionChanged -= OnTextViewCaretPositionChanged;
+                }
+            }
+        }
+
+        private void OnTextViewCaretPositionChanged(object sender, CaretPositionChangedEventArgs e)
+        {
+            var fireAndForget = UpdateDefinitionContext(e.TextView, e.NewPosition);
+        }
+
+        private async Task UpdateDefinitionContext(ITextView textView, CaretPosition caretPosition)
+        {
+            // Cancel any pending update for this view
+            _views[textView].Cancel();
+
+            // See if we moved somewhere else in a projection that we care about
+            var pointInRoslynSnapshot = caretPosition.Point.GetPoint(tb => tb.ContentType.IsOfType(ContentTypeNames.RoslynContentType), caretPosition.Affinity);
+            if (pointInRoslynSnapshot == null)
+            {
+                return;
+            }
+
+            // After a delay in case the caret moves again, find the symbol under the caret and update the context
+            var cancellationTokenSource = new CancellationTokenSource();
+            _views[textView] = cancellationTokenSource;
```

Comment:
Ah, the assumption is that this is foreground affinitized? Given it's saving you just the single marshal from the background thread to the foreground thread to call SetContext, what about just doing an old fashioned .ContinueWith instead?